src/data/error.rs

```diff
@@ -1,14 +1,59 @@
+use std::collections::VecDeque;
 use std::fmt;
-
 use thiserror::Error;
 
-use super::{Expr, Locatable, Location};
+use super::{Locatable, Location};
 
-pub type RecoverableResult<T = Expr, E = CompileError> = Result<T, (E, T)>;
+/// RecoverableResult is a type that represents a Result that can be recovered from.
+///
+/// See the [`Recover`] trait for more information.
+///
+/// [`Recover`]: trait.Recover.html
+pub type RecoverableResult<T, E = CompileError> = Result<T, (E, T)>;
 pub type CompileResult<T> = Result<T, CompileError>;
 pub type SemanticError = Locatable<String>;
 pub type CompileError = Locatable<Error>;
 
+/// ErrorHandler is a struct that hold errors generated by the compiler
+///
+/// An error handler is used because multiple errors may be generated by each
+/// part of the compiler, this cannot be represented well with Rust's normal
+/// `Result`.
+#[derive(Debug, Default, PartialEq)]
+pub(crate) struct ErrorHandler {
+    errors: VecDeque<CompileError>,
+}
+
+impl ErrorHandler {
+    /// Construct a new error handler.
+    pub(crate) fn new() -> ErrorHandler {
+        Default::default()
+    }
+
+    /// Add an error to the error handler.
+    pub(crate) fn push_err(&mut self, error: CompileError) {
```

Comment:
nit: can you change this to `push_back()`? We already know it's an error. See below for why `_back`.